### PR TITLE
Format positional arguments

### PR DESCRIPTION
**Pull request by cvwillegen**
This PR adds the possibility to use positional arguments in the printf() function, and in translation files, opening the possibility for translators to specify word order in translations. 

Ex: echo printf("My name is %2$s, %1$s", "Bram", "Moolenaar")

echo printf("My date of birth is: %1$02d/%2$02d/%3$04d", 13, 7, 2002)

echo printf("My date of birth is: %3$04d/%2$02d/%1$02d", 13, 7, 2002)

echo printf("My date of birth is: %3$0*1$d/%4$0*1$d/%5$0*2$d", 2, 4, 13, 7, 2002)

### CODE REVIEW

**Comment by cvwillegen** [comment]
Fixes #10577

**Comment by cvwillegen** [comment]
Hi,

On Tue, May 23, 2023 at 9:21 PM Restorer ***@***.***> wrote:
> @cvwillegen
> If I understood you correctly, then the built-in gettext tools can control the order of arguments during translation.
> Described here in this section of the gettext documentation.

Yes, that is the way to specify it in the .po files, but the Vim
runtime needs to be able to format strings that are written that way,
and it wasn't (isn't) able to. This PR adds that functionality so that
translators are more free to translate to correct word order.

Christ van Willegen

**Comment by yegappan** [comment]
Can you please update and rebase this PR to the latest?

**Comment by chrisbra** [comment]
can someone please try this out?

File: src/globals.h [hunk 1]
```diff
@@ -1674,6 +1674,19 @@ EXTERN int	cmdwin_result INIT(= 0); // result of cmdline window or 0
 
 EXTERN char_u no_lines_msg[]	INIT(= N_("--No lines in buffer--"));
 
+EXTERN char typename_unknown[]	INIT(= N_("unknown"));
+EXTERN char typename_int[]	INIT(= N_("int"));
+EXTERN char typename_longint[]	INIT(= N_("long int"));
+EXTERN char typename_longlongint[]	INIT(= N_("long long int"));
+EXTERN char typename_unsignedint[]	INIT(= N_("unsigned int"));
```
Comment (RestorerZ) [inline src/globals.h:1681]: I'm not sure if the format specifier type needs to be translated in the error message. Types of the C language are most often written int, long, etc.

**Comment by chrisbra** [comment]
@cvwillegen 
This PR introduced 2 warnings in Coverity. Can you please have a look at it:

```
Hi,

Please find the latest report on new defect(s) introduced to vim found with Coverity Scan.

2 new defect(s) introduced to vim found with Coverity Scan.


New defect(s) Reported-by: Coverity Scan
Showing 2 of 2 defect(s)


** CID 1539915:  API usage errors  (VARARGS)
/src/strings.c: 2881 in skip_to_arg()


________________________________________________________________________________________________________
*** CID 1539915:  API usage errors  (VARARGS)
/src/strings.c: 2881 in skip_to_arg()
2875         // Because we know that after we return from this call,
2876         // a va_arg() call is made, we can pre-emptively
2877         // increment the current argument index.
2878         ++*arg_cur;
2879         ++*arg_idx;
2880     

                CID 1539915:  API usage errors  (VARARGS)
                "va_end" was not called for "*ap".

2881         return;
2882     }
2883     
2884         int
2885     vim_vsnprintf_typval(
2886         char    *str,

** CID 1539914:    (FORWARD_NULL)


________________________________________________________________________________________________________
*** CID 1539914:    (FORWARD_NULL)
/src/strings.c: 3309 in vim_vsnprintf_typval()
3303                 // signed
3304                 switch (length_modifier)
3305                 {
3306                 case '\0':
3307                 case 'h':
3308                     // char and short arguments are passed as int.

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3309                     int_arg =
3310     # if defined(FEAT_EVAL)
3311                         tvs != NULL ? tv_nr(tvs, &arg_idx) :
3312     # endif
3313                             (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
3314                             va_arg(ap, int));
/src/strings.c: 3583 in vim_vsnprintf_typval()
3577                 double    f;
3578                 double    abs_f;
3579                 char    format[40];
3580                 int        l;
3581                 int        remove_trailing_zeroes = FALSE;
3582     

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3583                 f =
3584     # if defined(FEAT_EVAL)
3585                 tvs != NULL ? tv_float(tvs, &arg_idx) :
3586     # endif
3587                     (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
3588                     va_arg(ap, double));
/src/strings.c: 3290 in vim_vsnprintf_typval()
3284     
3285                 if (ptr_arg != NULL)
3286                     arg_sign = 1;
3287                 }
3288                 else if (fmt_spec == 'b' || fmt_spec == 'B')
3289                 {

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3290                 bin_arg =
3291     # if defined(FEAT_EVAL)
3292                         tvs != NULL ?
3293                            (uvarnumber_T)tv_nr(tvs, &arg_idx) :
3294     # endif
3295                         (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
/src/strings.c: 3166 in vim_vsnprintf_typval()
3160                 break;
3161     
3162             case 'c':
3163                 {
3164                 int j;
3165     

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3166                 j =
3167     # if defined(FEAT_EVAL)
3168                     tvs != NULL ? tv_nr(tvs, &arg_idx) :
3169     # endif
3170                     (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
3171                     va_arg(ap, int));
/src/strings.c: 3277 in vim_vsnprintf_typval()
3271                 // conversion
3272                 void *ptr_arg = NULL;
3273     
3274                 if (fmt_spec == 'p')
3275                 {
3276                 length_modifier = '\0';

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3277                 ptr_arg =
3278     # if defined(FEAT_EVAL)
3279                      tvs != NULL ? (void *)tv_str(tvs, &arg_idx,
3280                                         NULL) :
3281     # endif
3282                         (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
/src/strings.c: 3033 in vim_vsnprintf_typval()
3027                 uj = 10 * uj + (unsigned int)(*p++ - '0');
3028                 arg_idx = uj;
3029     
3030                 ++p;
3031             }
3032     

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3033             j =
3034     # if defined(FEAT_EVAL)
3035                 tvs != NULL ? tv_nr(tvs, &arg_idx) :
3036     # endif
3037                 (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
3038                 va_arg(ap, int));
/src/strings.c: 3181 in vim_vsnprintf_typval()
3175                 str_arg = (char *)&uchar_arg;
3176                 break;
3177                 }
3178     
3179             case 's':
3180             case 'S':

                CID 1539914:    (FORWARD_NULL)
                Passing null pointer "ap_types" to "skip_to_arg", which dereferences it.

3181                 str_arg =
3182     # if defined(FEAT_EVAL)
3183                     tvs != NULL ? tv_str(tvs, &arg_idx, &tofree) :
3184     # endif
3185                         (skip_to_arg(ap_types, ap_start, &ap, &arg_idx, &arg_cur),
3186                         va_arg(ap, char *));
```

File: src/strings.c [hunk 2]
```diff
@@ -2242,17 +2242,665 @@ vim_vsnprintf(
...
+    return OK;
+
+error:
+    vim_free(*ap_types);
+    *ap_types = NULL;
+    *num_posarg = 0;
+    return FAIL;
+}
+
+    void
```
Comment (yegappan) [inline src/strings.c:2791]: This function is used only within this file, so this function can be changed to a static function.

File: src/strings.c [hunk 3]
```diff
@@ -2242,17 +2242,665 @@ vim_vsnprintf(
...
+    va_list	ap_start,
+    va_list	*ap,
+    int		*arg_idx,
+    int		*arg_cur)
+{
+    int		arg_min = 0;
+
+    if (*arg_cur + 1 == *arg_idx)
+    {
+	++*arg_cur;
```
Comment (yegappan) [inline src/strings.c:2803]: Can you add a comment here describing why this is needed (similar to the ones below)?

**Comment by zeertzjq** [comment]
> I will check for NULL in the function and make it an internal error, because this shouldn't happen.

I think it can happen, but in that case it is never dereferenced.

**Comment by zeertzjq** [comment]
For example, it can happen when the first format specifier is `%c`.

**Comment by chrisbra** [comment]
@cvwillegen any updates here?

**Comment by cvwillegen** [comment]
I'll get to it after my holiday...

**Comment by chrisbra** [comment]
sorry, no problem, please have your time. I just wanted not to get this lost.

**Comment by chrisbra** [comment]
/remind me about it in 1 week

**Comment by cvwillegen** [comment]
> For example, it can happen at the first format specifier.

I looked at the code again. My conclusion is:

When there are no positional arguments specified, ap_types remains NULL. ap_types is, indeed, dereferenced in skip_to_arg() without checking for NULL, but _only if_ the argument being requested differs from the last one + 1, and that can only happen if there _are_ positional arguments. So, the skipping code is not executed.

If ap_types is NULL when positional arguments _are_ specified, then the parse_fmt_types() function has failed allocating memory somewhere and it returns FAIL, and in this case the rest of the code is not executed.

I _did_ add an internal error on the dereferencing (should not happen), but that's basically to shut up the warning :-) If it does happen, at least we know the format string an index where this went wrong.

**Pull request closed without merging**